### DISCUSSION TITLE
Thonny 4.1.4 (and other versions) installation problems

### DISCUSSION THREAD

**Discussion by BitHacker49** [thonny/thonny#3249] [General]
Hi, I'm new to Thonny and am experiencing unsurmountable problems with installing it on a Windows 11 computer.

The installation process pretends to work smoothly, proudly closing with "Great success...!", but the result is a mess. As soon as I try to start Thonny, the error message below is its response. This happened to other versions (I think 3.1.2 and 3.2.7) before. When I deinstall Thonny, tons of key entries containing "Thonny" are left in the Windows registry. I managed to delete them manually but I hate to fiddle around with the registry, knowing close to nothing about it.

Any idea what's wrong with my computer?

![image](https://github.com/thonny/thonny/assets/174866332/13320027-8449-4157-873b-bcbda9b7982a)

**Comment by aivarannamaa** [comment]
It looks like Thonny is having troubles querying your Windows username. I'll try using another approach in the next Thonny version.

Do you happen to have another Python installation in your system? If yes, then please tell what do you get with following script:

```python
import os.path
print(os.path.expanduser("~"))

import getpass
print(getpass.getuser())
```

> When I deinstall Thonny, tons of key entries containing "Thonny" are left in the Windows registry.

Could you give me some examples, which keys remain in the registry?

**Reply by Hermann12** [reply]
I think even the portable version writes in regedit. I found something like _python3.bat.FriendlyAppName_ ?

**Reply by BitHacker49** [reply]
> Do you happen to have another Python installation in your system? If yes, then please tell what do you get with following script:

With Python 3.10.6 I get:

> C:\Users\chris

**Comment by BitHacker49** [comment]
Thanks a lot for your response and sorry for my late answer.

The full truth is: All I reported above holds for my "normal" Windows user "chris" as you can see from the above screenshot. And that's where I finally want Thonny to run. Meanwhile - more as an act of desperation - I managed to install Thonny 3.2.7 under a different user account on the same physical computer (with booth users hooked to my single Microsoft account). 

This configuration somehow works but is a lousy solution because it forces me to switch users every now and then. (If you wonder about the old Thonny version number: that one was proposed by the freenove ESP32 kit I'm currently working with). 

So what I'm finally aiming at is: installing Thonny 4.1.4 under user "chris".

Concerning your question re: left-over entries in the registry I'll be checking out and respond soon.

**Comment by BitHacker49** [comment]
... and here we go:

> Computer\HKEY_USERS\S-1-5-21-871023383-4228313467-150262391-1002\Software\Microsoft\Windows\CurrentVersion\CloudStore\Store\DefaultAccount\Cloud\{a004fe6f-f24b-429a-bf23-f7b7ff8f44ee}$windows.data.apps.appleveltileinfo$appleveltilelist\windows.data.apps.appleveltileinfo$w~{7c5a40ef-a0fb-4bfc-874a-c0f2e0b9fa8e}thonnythonny.exe
> 
> Computer\HKEY_USERS\S-1-5-21-871023383-4228313467-150262391-1002\Software\Microsoft\Windows\CurrentVersion\CloudStore\Store\DefaultAccount\Cloud\{a004fe6f-f24b-429a-bf23-f7b7ff8f44ee}$windows.data.apps.appmetadata$appmetadatalist\windows.data.apps.appmetadata$thonny_is1
> 
> Computer\HKEY_USERS\S-1-5-21-871023383-4228313467-150262391-1002\Software\Microsoft\Windows\CurrentVersion\CloudStore\Store\DefaultAccount\Current\{a004fe6f-f24b-429a-bf23-f7b7ff8f44ee}$windows.data.apps.appleveltileinfo$appleveltilelist\windows.data.apps.appleveltileinfo$w~{7c5a40ef-a0fb-4bfc-874a-c0f2e0b9fa8e}thonnythonny.exe
> 
> Computer\HKEY_USERS\S-1-5-21-871023383-4228313467-150262391-1002\Software\Microsoft\Windows\CurrentVersion\CloudStore\Store\DefaultAccount\Current\{a004fe6f-f24b-429a-bf23-f7b7ff8f44ee}$windows.data.apps.appleveltileinfo$appleveltilelist\windows.data.apps.appleveltileinfo$w~cuserschrisappdatalocalprogramsthonnythonny.exe
> 
> Computer\HKEY_USERS\S-1-5-21-871023383-4228313467-150262391-1002\Software\Microsoft\Windows\CurrentVersion\CloudStore\Store\DefaultAccount\Current\{a004fe6f-f24b-429a-bf23-f7b7ff8f44ee}$windows.data.apps.appmetadata$appmetadatalist\windows.data.apps.appmetadata$thonny_is1
> 
> Computer\HKEY_USERS\S-1-5-21-871023383-4228313467-150262391-1002\Software\Microsoft\Windows\CurrentVersion\Search\JumplistData --- C:\Users\chris\AppData\Local\Programs\Thonny\thonny.exe
> 
> Computer\HKEY_USERS\S-1-5-21-871023383-4228313467-150262391-1002\Software\Microsoft\Windows\CurrentVersion\UFH\SHC --- 10 --- 
> C:\Users\chris\AppData\Roaming\Microsoft\Windows\Start Menu\Programs\Thonny\Thonny.lnk
> C:\Users\chris\AppData\Local\Programs\Thonny\thonny.exe


...plus 18 further key entries


Note: The "---" are inserted by myself

**Comment by BitHacker49** [comment]
Tried to install on my old Win10 PC both Thonny 4.1.4 and (after de-installing again) 3.2.7. Same result: Thonny doesn't start. 

Considering now to find something else ... :-(

**Comment by BitHacker49** [comment]
Hi, Aivaar,

I couldn't help playing around with this issue a bit further. I think your above assumption ("It looks like Thonny is having troubles querying your Windows username.") may be correct.

My normal account (were I do 99,9% of my work and where I'm having the trouble with Thonny) is a Microsoft account defined by two words (1st name and familiy name), separated by a blank. In the directory paths of my computers, only the first name appears (C:\users\<firstname>\...).

On both computers, I had already configured a second account the name of which is just a single word. In both of these, Thonny is starting successfully (version 3.2.7 on my desktop PC under Windows 11, version 4.1.4 on my 10 years old Notebook under Win 10). 

In the accounts working correctly, I find an empty directory C:\users\<alt_name>\AppData\Local\thonny-<name>; in the bad Microsoft accounts these directories are missing.

I should mention that I also installed Thonny 3.2.7 on my grandson's Notebook and it works there as well. (As far as I remember, his account name is a one-worder either.)

Until you've found a solution (thanks for trying!) I'll work with Thonny installed in those never-used side accounts. However, if all the above makes sense there should be no reason for not upgrading Thonny 3.2.7 to 4.1.4  on the Notebook.

One possible workaround could be to change my two-word Microsoft password into one word. However, when looking at where this can be changed in the Micosoft account information, they ask explicitly for first and family name. No idea what happens elsewhere when trying to leave the family name out. Windows is so spooky...

**Comment by aivarannamaa** [comment]
@BitHacker49, sorry for the late reply!

Which of the statements gave you "C:\Users\chris" -- `os.path.expanduser("~")` or `getpass.getuser()`?

**Comment by BitHacker49** [comment]
Hi Aivar,

I get this with both statements, see below:

```
Python 3.10.6 (tags/v3.10.6:9c7b4bd, Aug  1 2022, 21:53:49) [MSC v.1932 64 bit (AMD64)] on win32
Type "help", "copyright", "credits" or "license" for more information.
>>> import os.path
>>> print(os.path.expanduser("~"))
C:\Users\chris

>>> import getpass
>>> print(getpass.getuser())
C:\Users\chris
>>>

```
Same for the newer Python 3.12.4.

**Edit**: This all holds for my primary user account on my Windows 11 PC, where I keep having the problems to install Thonny.

I have a secondary account on that same PC where I succeeded to install Thonny weeks ago and which I'm using intensively since. The result there is (true account name replaced by the placeholder <2ndAccount>):

```
Python 3.12.5 (tags/v3.12.5:ff3bc82, Aug  6 2024, 20:45:27) [MSC v.1940 64 bit (AMD64)] on win32
Type "help", "copyright", "credits" or "license" for more information.
>>> import os.path
>>> print(os.path.expanduser("~"))
C:\Users\<2ndAccount>
>>>
>>> import getpass
>>> print(getpass.getuser())
<2ndAccount>
>>>
```

The last line is clearly different from the behaviour of the primary (failing) account.

Thanks for trying to help me; I'm longing to getting rid of that secondary account :-)

**Comment by t2solve** [comment]
thanks for the wonderful IDE !
i had the same issue, my windows os was generating a weird user name for the  AppData folders 
the username was including path chancing expression like: ".../" and so on
I found the issue at 'thonny/\_\_init\_\_.py' file:
I just change line 192:
```python
    ipc_dir = os.path.join(base_dir, "thonny-%s" % username)
```
to
```python
    ipc_dir = os.path.join(base_dir, "thonnyWonderwoman")
```
and its works !

**Reply by BitHacker49** [reply]
Thanks, this seems to solve it for me, too!

Just to make sure I understand it right:
- Since I deleted Thonny from my standard Windows account after all the trouble reported above, I reinstalled now Thonny once again. However, meanwhile the released version ist 4.1.6 rather than 4.1.4. So I chose 4.1.6. 
- Without your amendment, also Thonny 4.1.6 crashed when starting it after fresh install.
- Then I followed your proposal:
- On my Computer, Thonny is installed in **\...\AppData\Local\Programs\Thonny**
- In this Thonny directory there is no file **__init__.py**
- So I looked at the error message I was getting so far when starting Thonny after fresh installation. There were problems inside an __init__.py which resides at \...\AppData\Local\Programs\Thonny\ **Lib\site-packages\thonny**\
- There ist indeed an __init__.py, line 192 of which shows exactly what you described.
- since I'm not Wonderwoman :-) I changed that line to ipc_dir = os.path.join(base_dir, "thonny-myuser") where myuser is the Windows account name under which I'm doing all this.

Is that correct? (Well, at least it seems to work.)

Anyhow - many thanks for your effort! This helps me a lot.

Would be great if Aivar could fix this at some point in time because I agree: Thonny is meanwhile a wonderful IDE, quite simple and handy.

**Comment by t2solve** [comment]
thanks for the feedback @BitHacker49  i created fix #3410 and pull request do might fix this in the future, with a more general approach, not everybody could be wonderwoman

**Comment by aivarannamaa** [comment]
Fixed in #3411, will be available in Thonny 5.0
